mlos_core/mlos_core/tests/optimizers/optimizer_metadata_test.py

```diff
@@ -0,0 +1,99 @@
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+#
+"""
+Tests for Optimizers using Metadata.
+"""
+
+from typing import Callable
+
+import logging
+import pytest
+
+import pandas as pd
+import ConfigSpace as CS
+
+from smac import MultiFidelityFacade as MFFacade
+from smac.intensifier.successive_halving import SuccessiveHalving
+
+from mlos_core.optimizers import (
+    OptimizerType, OptimizerFactory, BaseOptimizer)
+from mlos_core.tests import SEED
+
+_LOG = logging.getLogger(__name__)
+_LOG.setLevel(logging.DEBUG)
+
+
+def smac_verify_best(metadata: pd.DataFrame) -> bool:
+    """
+    Function to verify if the metadata used by SMAC is in a legal state
+
+    Parameters
+    ----------
+    metadata : pd.DataFrame
+        metadata returned by SMAC
+
+    Returns
+    -------
+    bool
+        if the metadata that is returned is valid
+    """
+    max_budget = metadata["budget"].max()
+    if isinstance(max_budget, float):
+        return max_budget == 9
+    return False
+
+
+@pytest.mark.parametrize(('optimizer_type', 'verify', 'kwargs'), [
+    # Enumerate all supported Optimizers
+    *[(member, verify, {"seed": SEED, "facade": MFFacade, "intensifier": SuccessiveHalving, "min_budget": 1, "max_budget": 9})
+      for member, verify in [(OptimizerType.SMAC, smac_verify_best)]],
```

Comment:
This is a pretty weird pattern.  If we wanted to add more types in the future those kwargs wouldn't work.

What we've done elsewhere is
1. retain the loop over `OptimizerType` in order to make sure that all optimizers get a test added (or explicitly excluded) whenever we add new optimizer backends.
2. When necessary, within the body of the test set additional parameters in a switch statement for a given optimizer type. e.g.,
```python
if optimizer_type == OptimizerType.SMAC:
  verifier = smac_verify_best
elif optimizer_type == OptimizerType.FLAML:
  pytest.skip("TODO: FLAML Optimizer does not yet support metadata")  # though even here, I think we should test *something*
else:
  raise NotImplementedError(f"Missing test handler for OptimizerType {optimizer_type}")

Reply:
I have moved the kwargs into the tuple which fixes the kwargs problem.

I find switching over unsupported optimizers messy since it means, in this case, we should skip the entire test as you have above, _unless_ we are testing SMAC. There are other places where we test to make sure that using metadata throws exceptions, so without metadata support there is nothing to test.

Reply:
But when we add that functionality to another optimizer it will be very easy to forget to enable related testing support, so I'd rather add that now as it will be easier to search for what we need to enable.

Reply:
Won't it be the same either way? One requires adding a line, the other requires moving a line. I make this more similar to other tests, but I don't think it changes much either way.